src/FSharpPlus/DList.fs

```diff
@@ -1,44 +1,151 @@
-    let toSeq  (DList x)   = x [] :> seq<_>
-    let ofList source = source |> List.append |> DList
-    let ofSeq  source = source |> Seq.toList |> List.append |> DList
-    let append (DList x) (DList y) = DList (x << y)
-    let singleton x = DList (fun r -> x::r)
-    let cons x (DList f) = (DList (fun l -> x::(f l)))
-    let snoc (DList f) x = (DList (fun l -> f (x::l)))
-    let fold f x = List.fold f x << toList
-    let map f (DList x) = x [] |> List.map f |> List.append |> DList
-    let concat x = List.fold append empty x
-    let join (DList f) = concat (f [])
+
+    let empty<'a>             = DList<'a> (0, Nil)
+    let toList (x:DList<_>)   = x.Walk [] |> Seq.toList
+    let toSeq  (x:DList<_>)   = x.Walk []
+    let ofSeq  source = 
+        DList(Seq.length source, (Seq.fold (fun state x ->
+            match state with 
+            | Nil -> Unit x
+            | Unit _ -> Join(state, Unit x)
+            | Join(_,_) as xs -> Join(state, Unit x)) Nil source))
+    let ofList source = 
+        DList(List.length source, (List.fold (fun state x ->
+            match state with 
+            | Nil -> Unit x
+            | Unit _ -> Join(state, Unit x)
+            | Join(_,_) as xs -> Join(state, Unit x)) Nil source))
+
+    let append (left: DList<_>) (right: DList<_>) = 
+        let len =left.Length + right.Length
+        let data= DListData.append left.dc right.dc
+        DList(len, data)
+
+    let singleton x = DList (1, Unit x)
+    let cons hd (f:DList<_>) =
+        match f.dc with
+        | Nil -> DList (1, (Unit hd))
+        | _ ->  DList ((f.Length + 1), Join(Unit hd, f.dc) )
+    let snoc (f:DList<_>) x = DList( (f.Length + 1), DListData.append (f.dc) (Unit x) )
+    let fold f x = DList<_>.Fold'(f x)
+    let map f (x:DList<_>) = DList<_>.FoldBack' (cons << f ) x empty
+    let concat x = DList<_>.Fold' append empty x
+    let join (f:DList<DList<_>>) = concat f
     let ap f x = join <| map (fun y -> map ((|>) y) f) x
-    let bind m k = concat << List.map k << toList <| m
+    let bind m k = DList<_>.FoldBack' (append << k) empty m
 
 open DList
 
 type DList<'T> with
     
-    static member get_Zero = DList id
-    static member (+) (DList x, DList y) = DList (x << y)
+    static member get_Zero = DList( 0, Nil)
+    static member (+) (x:DList<_>, y:DList<_>) = DList.append x y
 
-    static member get_Empty = DList id
-    static member Append (DList x, DList y) = DList (x << y)
+    static member get_Empty = DList( 0, Nil)
+    static member Append (x:DList<_>, y:DList<_>) = DList.append x y
     
     static member ToSeq  x = toSeq  x
     static member ToList x = toList x
     static member OfSeq  x = ofSeq  x
     static member OfList x = ofList x
     static member Fold (x, f, z, [<Optional>]_impl) = fold f x z    
 
-    static member Return x = DList (fun r -> x::r)
+    static member Return x = DList (1, x)
     static member Map (x, f) = map f x
     static member (<*>) (f, x) = ap f x
```

Comment:
Although you opened the module DList you can write these operations qualified (ie DList.map instead of map), it will makes thing clear.